perf/change_point_detector.go

```diff
@@ -0,0 +1,79 @@
+package perf
+
+import (
+	"bytes"
+	"encoding/json"
+	"github.com/evergreen-ci/cedar/util"
+	"github.com/evergreen-ci/gimlet"
+	"github.com/pkg/errors"
+	"net/http"
+)
+
+type ChangeDetector interface {
+	DetectChanges([]float64) ([]ChangePoint, error)
+}
+
+type ChangePoint struct {
+	Index     int
+	Algorithm Algorithm
+}
+
+type Algorithm struct {
+	Name          string
+	Version       string
+	Configuration map[string]float64
+}
+
+type signalProcessingClient struct {
+	token   string
+	baseURL string
+}
+
+func NewChangeDetector(baseURL, token string) ChangeDetector {
+	return &signalProcessingClient{token: token, baseURL: baseURL}
+}
+
+func (spc *signalProcessingClient) DetectChanges(series []float64) ([]ChangePoint, error) {
+	changePoints := &struct {
+		ChangePoints []ChangePoint `json:"changePoints"`
+	}{}
+
+	if err := spc.doRequest(http.MethodPost, "change_points/detect", series, changePoints); err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	return changePoints.ChangePoints, nil
+}
+
+func (spc *signalProcessingClient) doRequest(method, route string, in, out interface{}) error {
+	body, err := json.Marshal(in)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	client := util.GetHTTPClient()
+	defer util.PutHTTPClient(client)
+
+	req, err := http.NewRequest(method, route, bytes.NewBuffer(body))
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	req.Header.Add("Authorization", "Bearer "+spc.token)
+	req.Header.Add("Content-Type", "application/json")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return errors.New("Failed to detect changes in metric data, status: "+string(resp.StatusCode))
```

Comment:
use errors.Errorf() rather than string concatenation to format the error string. 

there's also http.StatusText() to get the thing you want. I'm not sure that casting an int to a string does exactly what you want. 

are there status codes in the 200 or 300 range that we should handle? are there codes in the 500 range that mean that we should retry with backoff? this question also applies to the error condition on line 66. 

There is some existing infrastructure in evergreen that we use to make each implementation of retries with backoff better, but you should decide what the conditions that are retriable and then use those tools, which I'm happy to help with.

Reply:
I've added this one for the retrying stuff and error handling story: https://jira.mongodb.org/browse/MAKE-1050

We will obviously do this before this goes operational, I'd just like to divorce it from this ticket since it can be framed as an improvement to the naive client (which believes nothing will go wrong, but shares the same interface with a client that gracefully handles recoverable errors).